### PR TITLE
Speed up GroupMax for large group sizes

### PR DESCRIPTION
**Pull request by kangshiyin**
New version uses multiple threads per group to speed up the max reduction.

~4x faster for large group sizes (>10).

```
New: For CuMatrix::GroupMax<float>, for dim = 16, speed was 0.0190836 gigaflops.
Old: For CuMatrix::GroupMax<float>, for dim = 16, speed was 0.0193129 gigaflops.
New: For CuMatrix::GroupMax<float>, for dim = 32, speed was 0.0791846 gigaflops.
Old: For CuMatrix::GroupMax<float>, for dim = 32, speed was 0.0768508 gigaflops.
New: For CuMatrix::GroupMax<float>, for dim = 64, speed was 0.311131 gigaflops.
Old: For CuMatrix::GroupMax<float>, for dim = 64, speed was 0.299519 gigaflops.
New: For CuMatrix::GroupMax<float>, for dim = 128, speed was 1.13589 gigaflops.
Old: For CuMatrix::GroupMax<float>, for dim = 128, speed was 1.14847 gigaflops.
New: For CuMatrix::GroupMax<float>, for dim = 256, speed was 4.22264 gigaflops.
Old: For CuMatrix::GroupMax<float>, for dim = 256, speed was 3.92072 gigaflops.
New: For CuMatrix::GroupMax<float>, for dim = 512, speed was 12.2629 gigaflops.
Old: For CuMatrix::GroupMax<float>, for dim = 512, speed was 10.0812 gigaflops.
New: For CuMatrix::GroupMax<float>, for dim = 1024, speed was 21.6979 gigaflops.
Old: For CuMatrix::GroupMax<float>, for dim = 1024, speed was 16.5123 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 16, speed was 0.0188551 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 16, speed was 0.0163827 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 32, speed was 0.0701613 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 32, speed was 0.0620238 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 64, speed was 0.271106 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 64, speed was 0.215268 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 128, speed was 0.931745 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 128, speed was 0.723582 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 256, speed was 3.53189 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 256, speed was 1.9751 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 512, speed was 9.95109 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 512, speed was 3.91183 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 1024, speed was 17.2099 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<float>, for dim = 1024, speed was 4.92671 gigaflops.
New: For CuMatrix::GroupMax<double>, for dim = 16, speed was 0.0199497 gigaflops.
Old: For CuMatrix::GroupMax<double>, for dim = 16, speed was 0.0148693 gigaflops.
New: For CuMatrix::GroupMax<double>, for dim = 32, speed was 0.079538 gigaflops.
Old: For CuMatrix::GroupMax<double>, for dim = 32, speed was 0.0718237 gigaflops.
New: For CuMatrix::GroupMax<double>, for dim = 64, speed was 0.314509 gigaflops.
Old: For CuMatrix::GroupMax<double>, for dim = 64, speed was 0.237838 gigaflops.
New: For CuMatrix::GroupMax<double>, for dim = 128, speed was 1.08104 gigaflops.
Old: For CuMatrix::GroupMax<double>, for dim = 128, speed was 0.788395 gigaflops.
New: For CuMatrix::GroupMax<double>, for dim = 256, speed was 3.7741 gigaflops.
Old: For CuMatrix::GroupMax<double>, for dim = 256, speed was 2.87856 gigaflops.
New: For CuMatrix::GroupMax<double>, for dim = 512, speed was 8.65988 gigaflops.
Old: For CuMatrix::GroupMax<double>, for dim = 512, speed was 5.87111 gigaflops.
New: For CuMatrix::GroupMax<double>, for dim = 1024, speed was 14.0373 gigaflops.
Old: For CuMatrix::GroupMax<double>, for dim = 1024, speed was 8.88655 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 16, speed was 0.0174585 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 16, speed was 0.0136057 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 32, speed was 0.0694617 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 32, speed was 0.0500527 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 64, speed was 0.265809 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 64, speed was 0.177945 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 128, speed was 0.973417 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 128, speed was 0.588654 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 256, speed was 3.43166 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 256, speed was 1.57864 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 512, speed was 8.26032 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 512, speed was 3.14173 gigaflops.
New: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 1024, speed was 12.1338 gigaflops.
Old: For CuMatrix::GroupMax (all group sizes)<double>, for dim = 1024, speed was 3.05406 gigaflops.
```

### CODE REVIEW

File: src/cudamatrix/cu-matrix.cc [hunk 1]
```diff
@@ -1309,11 +1309,22 @@ void CuMatrixBase<Real>::GroupMax(const CuMatrixBase<Real> &src) {
...
-    dim3 dimGrid(n_blocks(NumCols(), CU2DBLOCK),
-                 n_blocks(NumRows(), CU2DBLOCK));
-    cuda_group_max(dimGrid, dimBlock, this->data_, src.data_,
-                   this->Dim(), src.Stride(), group_size);
+    // One thread block per row.
+    // Use 2D block for small group size to simplify the calculation
+    // Use 1D block for large group size
+    //   group size: 1 2 3 4 5 6 7 .. 12 13 .. 24 25 .. 48 49 .. 96 97 .. 192 193 .. 384 385 ...
+    // ragular_size: 1 1 1 2 2 2 4 ..  4  8 ..  8 16 .. 16 32 .. 32 64 ..  64 128 .. 128 256 ...
+    int ragular_size = CU1DBLOCK;
```
Comment (danpovey) [inline src/cudamatrix/cu-matrix.cc]: perhaps you mean 'regular_size'.  But even then, I don't understand why you're using this word at all- perhaps it's a mistranslation?  What is the concept behind it?

Comment (kangshiyin) [inline src/cudamatrix/cu-matrix.cc]: It's a typo. I was trying to say it should be a power of 2... 
I changed the name to `threads_per_group`

File: src/cudamatrix/cu-matrix.cc [hunk 2]
```diff
@@ -1309,11 +1309,23 @@ void CuMatrixBase<Real>::GroupMax(const CuMatrixBase<Real> &src) {
...
-                 n_blocks(NumRows(), CU2DBLOCK));
-    cuda_group_max(dimGrid, dimBlock, this->data_, src.data_,
-                   this->Dim(), src.Stride(), group_size);
+    // One thread block per row.
+    // Use 2D block for small group size to simplify the calculation.
+    // Each group is reduced by threads_per_group threads.
+    // threads_per_group should be a power of 2 for fast tree reduction.
+    //        group size: 1 2 3 4 5 6 7 .. 12 13 .. 24 25 .. 48 ...
+    // threads_per_group: 1 1 1 2 2 2 4 ..  4  8 ..  8 16 .. 16 ...
+    int threads_per_group = CU1DBLOCK;
```
Comment (kangshiyin) [inline src/cudamatrix/cu-matrix.cc]: I was trying to say the size should be a power of 2 ...
I've changed the name to `threads_per_group` and added a comment to explain.

**Comment by danpovey** [comment]
Thanks!!  Merging.
BTW, since you're optimizing kernels, see if you can get a setup with the chain models running so you can figure out which kernels it's taking a lot of time in.

**Comment by kangshiyin** [comment]
OK, I will see what I can do with it.

**Pull request merged**